bug: No such file or directory error on mac

#### env
os: sonoma 14.6.1
arch: x86-64

#### steps
1. `cargo install cargo-machete --version 0.6.2`
2. `git clone https://github.com/systemaccounting/mxfactorial`
3. `cd mxfactorial/services/graphql`
4. `cargo machete`

#### observed
```
Analyzing dependencies of crates in this directory...
error when handling ./Cargo.toml: No such file or directory (os error 2)
cargo-machete didn't find any unused dependencies in .. Good job!
Done!
```

#### expected
no repro with `rm ~/.cargo/bin/cargo-machete && cargo install cargo-machete --version 0.6.1 && cargo machete`:
```
Analyzing dependencies of crates in this directory...
cargo-machete didn't find any unused dependencies in /private/tmp/mxfactorial/services/graphql. Good job!
Done!
```

#### release
you are permitted to clone a copy of the privately licensed https://github.com/systemaccounting/mxfactorial repo for debugging this issue

please push a [rust-version](https://doc.rust-lang.org/cargo/reference/rust-version.html)

`cargo build --release` errors on `rustc 1.80.1 (3f5fd8dd4 2024-08-06)` with:
```sh
   Compiling cargo-machete v0.7.0 (/Users/mx/code/cargo-machete)
error[E0599]: no method named `is_none_or` found for enum `std::option::Option` in the current scope
   --> src/search_unused.rs:184:18
    |
181 |               if dir_entry
    |  ________________-
182 | |                 .path()
183 | |                 .extension()
184 | |                 .is_none_or(|ext| ext.to_string_lossy() != "rs")
    | |_________________-^^^^^^^^^^
    |
help: there is a method `is_none` with a similar name, but with different arguments
   --> /Users/mx/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/option.rs:653:5
    |
653 |     pub const fn is_none(&self) -> bool {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0599`.
error: could not compile `cargo-machete` (bin "cargo-machete") due to 1 previous error
```

available to resume testing tomorrow

--- a/src/main.rs
+++ b/src/main.rs
@@ -177,7 +177,7 @@ fn run_machete() -> anyhow::Result<bool> {
                     }
 
                     Err(err) => {
-                        eprintln!("error when handling {}: {}", manifest_path.display(), err);
+                        eprintln!("error when handling {}: {:#}", manifest_path.display(), err);
                         None
                     }
                 },

--- a/src/search_unused.rs
+++ b/src/search_unused.rs
@@ -4,7 +4,7 @@ use grep::{
     regex::{RegexMatcher, RegexMatcherBuilder},
     searcher::{self, BinaryDetection, Searcher, SearcherBuilder, Sink},
 };
-use log::{debug, trace};
+use log::{debug, trace, warn};
 use rayon::prelude::*;
 use std::{
     collections::{BTreeMap, HashSet},
@@ -300,7 +300,14 @@ fn get_full_manifest(
     let mut ws_manifest_and_path = None;
     let mut workspace_ignored = vec![];
 
-    let mut dir_path = dir_path.to_path_buf();
+    // Canonicalize the path, so as to get the full "parenthood" of relative paths.
+    let mut dir_path = std::fs::canonicalize(dir_path).unwrap_or_else(|err| {
+        warn!("error when canonicalizing dir_path: {err}");
+        dir_path.to_owned()
+    });
+
+    // Try to find a workspace manifest, starting from the current directory, going up to the
+    // filesystem's root.
     while dir_path.pop() {
         let workspace_cargo_path = dir_path.join("Cargo.toml");
         if let Ok(workspace_manifest) =
@@ -936,4 +943,30 @@ fn test_ignore_deps_workspace_works() {
             assert_eq!(analysis.ignored_used, &["rand_core".to_string()]);
         },
     );
+}
+
+#[test]
+fn test_workspace_from_relative_path() {
+    // Ensure that finding a workspace using a relative path works.
+    use std::env::{current_dir, set_current_dir};
+
+    let prev_cwd = current_dir().unwrap();
+
+    set_current_dir(
+        PathBuf::from(TOP_LEVEL).join("./integration-tests/workspace-package/program/"),
+    )
+    .unwrap();
+
+    let path = Path::new("./Cargo.toml");
+    let analysis = find_unused(&path, UseCargoMetadata::No);
+
+    // Reset the current directory *before* running any other check.
+    set_current_dir(prev_cwd).unwrap();
+
+    let analysis = analysis
+        .expect("find_unused must return an Ok result")
+        .expect("no error during processing");
+
+    assert_eq!(analysis.unused, &["log".to_string()]);
+    assert!(analysis.ignored_used.is_empty());
 }

